sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -539,28 +538,26 @@ as we can guarantee we only need at most two buffers for any call to write (two
     private Mono<Void> uploadBlocks(long fileSize, ParallelTransferOptions parallelTransferOptions,
         BlobHttpHeaders headers, Map<String, String> metadata, AccessTier tier,
         BlobRequestConditions accessConditions, AsynchronousFileChannel channel, BlockBlobAsyncClient client) {
-        final ParallelTransferOptions finalParallelTransferOptions = (parallelTransferOptions == null)
-            ? new ParallelTransferOptions(null, null, null)
-            : new ParallelTransferOptions(parallelTransferOptions.getBlockSize(),
-            parallelTransferOptions.getNumBuffers(), parallelTransferOptions.getProgressReceiver());
+        final ParallelTransferOptions finalParallelTransferOptions =
+            ModelHelper.populateAndApplyDefaults(parallelTransferOptions);
         final BlobRequestConditions finalRequestConditions = (accessConditions == null)
-            ? new BlobRequestConditions()
-            : accessConditions;
-        ProgressReceiver progressReceiver = finalParallelTransferOptions.getProgressReceiver();
+            ? new BlobRequestConditions() : accessConditions;
+
 
         // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
         AtomicLong totalProgress = new AtomicLong();
         Lock progressLock = new ReentrantLock();
 
         final SortedMap<Long, String> blockIds = new TreeMap<>();
-        return Flux.fromIterable(sliceFile(fileSize, finalParallelTransferOptions.getBlockSize()))
+        return Flux.fromIterable(sliceFile(fileSize, finalParallelTransferOptions.getBlockSize(),
+            parallelTransferOptions))
```

Comment:
Are file size and block size not part of parallel transfer options?

Reply:
fileSize is not. We get it from the file. We take in the parallelTransferOptions (not the final ones) because we want to be able to check if the customer explicitly set a block size so we don't overwrite it if they did.